File: 20230604_Sunday/20230604.go
// Leetcode: 547. Number of Provinces.
// https://leetcode.com/problems/number-of-provinces/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 26 ms, faster than 36.21% of Go online submissions for Number of
// Provinces.
// Memory Usage: 6.8 MB, less than 65.43% of Go online submissions for Number of
// Provinces.
// 2023.06.05 Updated.
package main

import (
	"fmt"
)

var g [][]int // Graph.
var v []bool  // Visited.

func findCircleNum(isConnected [][]int) int {
	g = isConnected
	v = make([]bool, len(g))
	ans := 0 // Number of components.
	for i := 0; i < len(g); i += 1 {
		if true == v[i] {
			continue
		}
		bfs(i)
		ans += 1
	}
	return ans
}

func bfs(i int) {
	v[i] = true
	q := make([]int, 1)
	q[0] = i
	for 0 != len(q) {
		x := q[0]
		q = q[1:]
		for j, _ := range g {
			if v[j] || 0 == g[x][j] {
				continue
			}
			q = append(q, j)
			v[j] = true
		}
	}
}

File: 20230604_Sunday/20230604.rb
# Leetcode: 547. Number of Provinces.
# https://leetcode.com/problems/number-of-provinces/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 99 ms, faster than 78.57% of Ruby online submissions for Number of
# Provinces.
# Memory Usage: 214 MB, less than 33.33% of Ruby online submissions for Number
# of Provinces.
# 2023.06.04 Daily Challenge.
# @param {Integer[][]} is_connected
# @return {Integer}
def find_circle_num(is_connected)
  # Based on:
  # https://leetcode.com/problems/number-of-provinces/solution/
  @icn = is_connected # Is connected.
  @n = is_connected.size
  @nc = 0 # Number of components.
  @v = Array.new(@n,false) # Visited.
  (0...@n).each do |i|
    next if @v[i]
    @nc += 1
    bfs(i)
  end
  return @nc
end

def bfs(i)
  @v[i] = true
  q = []
  q.push(i)
  while !q.empty?
    node = q.shift()
    (0...@n).each do |j|
      next if @v[j]
      next if 0 == @icn[node][j]
      q.push(j)
      @v[j] = true
    end
  end
end
